src/controller.py

```diff
@@ -0,0 +1,585 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""nginx-ingress-integrator k8s resource controllers."""
+
+import abc
+import logging
+import typing
+
+import kubernetes.client
+
+from consts import CREATED_BY_LABEL
+from options import IngressOption
+
+logger = logging.getLogger(__name__)
+
+AnyResource = typing.TypeVar(  # pylint: disable=invalid-name
+    "AnyResource",
+    kubernetes.client.V1Endpoints,
+    kubernetes.client.V1EndpointSlice,
+    kubernetes.client.V1Service,
+    kubernetes.client.V1Ingress,
+)
+
+
+class ResourceController(abc.ABC, typing.Generic[AnyResource]):
```

Comment:
I was wondering if a `Protocol` might be more suited here than ABCs, since from what I understand, we are trying to define interfaces for the charm to use.

Reply:
Yes, the ABC and Protocol with explicitly declaring implementation are basically the same. It's more of a personal preference. Since we don't need implicit subtyping with structural typing, I just chose the ABC.

Reply:
Actually, now I think the generic syntax of Protocol is more elegant, change to Protocol with explicitly declaring implementation. Thanks for the idea!